### PR TITLE
Optional second stage population step.

### LINKED ISSUE #54: Provide a hook to perform 'secondary' render
It is not performant to populate list items with images when scrolling at speed. For this reason we must allow the user to provide a second populate callback that we call only when we think we have time (either scrolling slowly or not at all).

``` js
new FastList({
  populateItem: function(el, i) { ... },
  populateItemDetail: function(el, i) {
    // load images then wait for 'load' and fade in
  }
});
```

To prevent old images showing up in recycled items we need to provide an 'unpoplateItemDetail' hook to give the user a chance to reset the changes they made.

``` js
new FastList({
  ...

  unpopulateItemDetail: function(el, i) {
    // hide <img>s and reset src
  }
});
```

---

In `<gaia-fast-list>` list items should only ever contain one image. This means we can be more opinionated and provide even simpler interface to the user.

``` js
list.configure({
  getImageSrc: function(i) {
    return getImageSrcForItem(i);
  }
});
```

Behind the scenes we can take care of setting the given `src`, fading-in the image on `'load'` and then cleaning up. This work should all be done when app is 'relatively' idle (slow or stopped scrolling), so performance isn't as much of a concern as it would be with `populateItem()`.

@etiennesegonzac any thoughts to add?

### PR DESCRIPTION
**Pull request by etiennesegonzac**
Closes #54

### CODE REVIEW

**Comment by etiennesegonzac** [comment]
Not tested or anything, just a sneak peek.

File: examples/detail/app.js [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+    body.firstChild.data = record.metadata.artist;
+  },
+
+  populateItemDetail: function(el, i) {
+    var img = el.querySelector('img');
+    img.src = makeImage(i);
+    img.onload = function() {
+      setTimeout(function() {
+        img.classList.add('loaded');
+      });
```
Comment (wilsonpage) [inline examples/detail/app.js]: What's the `setTimeout` for, triggering the transition?

Comment (etiennesegonzac) [inline examples/detail/app.js]: maybe some leftover from before I figured out I needed to wait for the load, will try without tomorrow :)

File: fast-list.js [hunk 2]
```diff
@@ -569,6 +600,11 @@ function tryToPopulate(item, index, source, first) {
     return;
   }
 
+  if (source.populateItemDetail && item.dataset.detailPopulated !== 'true') {
+    source.populateItemDetail(item, index);
```
Comment (wilsonpage) [inline fast-list.js:614]: Is this being delayed until `idle`?

**Comment by wilsonpage** [comment]
I haven't yet got my head round how this works. Can you give a high level summary of the approach you've taken?

File: fast-list.js (hunk 2 above)
Comment (etiennesegonzac) [inline fast-list.js:614]: this one is for the "content coming late" use case.

**Comment by etiennesegonzac** [comment]
The idea is not to tie rendering to how fast the list is scrolling but how much lag there is.
Because of course we want to be more generous with loading images on a high-end device regardless of the speed.

This actually how the "fast scroll" mode was working all along but I'm trying to make things more explicit.

The best indicator we have of the lag is how much we moved between 2 handled scroll events. If we start busting the main thread we'll get less event and thus travel further between the 2 (since the actual scrolling is off the main thread).

So from this indicator we derive 2 booleans:
- are we particularly idle, we get a lot of events along the scroll, we can safely do the extra populateItemDetail work
- are we particularly busy, we're probably behind the actual scrolling position anyway, let's skip rendering altogether

File: fast-list.js (hunk 2 above)
Comment (wilsonpage) [inline fast-list.js:614]: Can you give more detail?

File: fast-list.js [hunk 3]
```diff
@@ -663,7 +699,7 @@ function schedulerShim() {
   var raf = window.requestAnimationFrame;
 
   return {
-    mutation: function(block) { return Promise.resolve(block()); },
+    mutation: function(block) { return Promise.resolve().then(block); },
```
Comment (wilsonpage) [inline fast-list.js:712]: Nice spot ;)

File: fast-list.js [hunk 4]
```diff
@@ -283,6 +302,17 @@ FastList.prototype = {
...
       var section = source.getSectionFor(i);
       placeItem(item, i, section, geo, source, reload);
+
+      if (!source.populateItemDetail) {
+        return;
+      }
+
+      if (item.dataset.detailPopulated === 'true' || !geo.idle) {
+        return;
+      }
```
Comment (wilsonpage) [inline fast-list.js]: I'm confused, if we abort `populateItemDetail()` here, does it get called later once the list has become `idle`?

File: examples/detail/app.js [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+  var left = (88 - width) / 2;
+  ctx.textBaseline = 'top';
+  ctx.fillStyle = '#000'
+  ctx.fillText(index, left - 3, top);
+  ctx.fillStyle = 'red'
+  ctx.fillText(index, left, top);
+  ctx.fillStyle = 'cyan'
+  ctx.fillText(index, left + 3, top);
+  return canvas.toDataURL('image/png');
+}
```
Comment (wilsonpage) [inline examples/detail/app.js:151]: Interested to see an example with normal images loaded through network too. Do you think this will make a difference?

Comment (etiennesegonzac) [inline examples/detail/app.js:151]: Network images will probably require something specific...
Requests could end at any time, and we don't have much control over the expensive steps.
We might need to make `populateItemDetail` return a promise in this case etc...

But I am eager to try out for the music use case in the component.

**Comment by etiennesegonzac** [comment]
@wilsonpage Updated!
The "late content support" isn't much more streamlined, but at least it's tested and documented a bit.

File: README.md [hunk 6]
```diff
@@ -40,6 +40,23 @@ var myList = new FastList({
   populateItem: function(el, index) { ... },
 
   /**
+   * Called when the ressources allows it to do more expensive rendering
```
Comment (wilsonpage) [inline README.md:43]: typo

File: fast-list.js [hunk 7]
```diff
@@ -173,28 +173,44 @@ FastList.prototype = {
...
+    var previousIdle = geo.idle;
+
+    if (!previousBusy  && moved > geo.viewportHeight * 2) {
+      previousBusy = true;
     }
 
-    if (fastScrolling && moved > 0 && moved < geo.viewportHeight * 0.5) {
-      fastScrolling = false;
+    if (previousBusy && moved > 0 && moved < geo.viewportHeight * 0.5) {
+      previousBusy = false;
```
Comment (wilsonpage) [inline fast-list.js]: NIT: This naming is a little confusing, as what we're really saying is `currentlyBusy = false`. Could we set `geo.busy = false` here instead and not change the value of `previouslyBusy`? Same applies for `previouslyIdle`.

File: fast-list.js [hunk 8]
```diff
@@ -283,6 +304,26 @@ FastList.prototype = {
 
       var section = source.getSectionFor(i);
       placeItem(item, i, section, geo, source, reload);
+
+      if (!source.populateItemDetail) {
+        return;
+      }
+
+
```
Comment (wilsonpage) [inline fast-list.js:306]: Remove this line

File: fast-list.js [hunk 9]
```diff
@@ -283,6 +304,26 @@ FastList.prototype = {
 
       var section = source.getSectionFor(i);
       placeItem(item, i, section, geo, source, reload);
+
+      if (!source.populateItemDetail) {
+        return;
+      }
```
Comment (wilsonpage) [inline fast-list.js]: Are we doing single-line ifs?

File: fast-list.js [hunk 10]
```diff
@@ -283,6 +304,26 @@ FastList.prototype = {
...
+
+
+      // Populating the item detail when we're not too busy scrolling
+      // Note: we're always settling back to an idle stage at some point where
+      // we do all the pending detail populations
+      if (!geo.idle) {
+        return;
+      }
```
Comment (wilsonpage) [inline fast-list.js]: Single-line?

File: fast-list.js [hunk 11]
```diff
@@ -283,6 +304,26 @@ FastList.prototype = {
...
+
+      // Detail already populated, skipping
+      if (item.dataset.detailPopulated === 'true') {
+        return;
+      }
```
Comment (wilsonpage) [inline fast-list.js]: Single-line?

File: test/fast-list-test.js [hunk 12]
```diff
@@ -349,4 +407,65 @@ suite('FastList >', function() {
...
+
+        trPromise.resolve();
+        return Promise.resolve().then(function() {
+          for (var i = 0; i < items.length; i++) {
+            var item = items[i];
+            assert.equal(item.style.transition, '');
+          }
+        });
+      });
+    });
```
Comment (wilsonpage) [inline test/fast-list-test.js:469]: Revealing new items with a transition is quite opinionated for the core library. What is the transition for? Perhaps plugin territory?

**Comment by wilsonpage** [comment]
Looks good few nits/questions. Will give a quick check on device, then ship it!!

File: test/fast-list-test.js (hunk 12 above)
Comment (etiennesegonzac) [inline test/fast-list-test.js:469]: yeah this part is "scheduled for plugin extraction", but it feels safer having the tests first :)

File: examples/detail/app.js (hunk 5 above)
Comment (etiennesegonzac) [inline examples/detail/app.js:151]: @wilsonpage what's an average size of the artwork in the music app?
We can make the canvas bigger here to make the performance impact more realistic.

Comment (wilsonpage) [inline examples/detail/app.js:151]: The default artwork is 600x600 jpegs

**Comment by wilsonpage** [comment]
Ship it! :D

File: fast-list.js (hunk 7 above)
Comment (etiennesegonzac) [inline fast-list.js]: yeah it was bugging me too, much better this way!

**Comment by etiennesegonzac** [comment]
Shipped :)

**Pull request merged**